cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -233,6 +235,282 @@ struct DivideChecked {
   }
 };
 
+struct Power {
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_signed_integer<T> Call(KernelContext* ctx, Arg0 left, Arg1 right) {
+    Arg0 result = 1;
+    if (left == 0) {
+      if (right < 0) {
+        ARROW_PREDICT_FALSE(MultiplyWithOverflow(2, std::numeric_limits<Arg0>::max(), &result));
+        ctx->SetStatus(Status::Invalid("divide by zero"));
+        return result;
+      } else if (right == 0) {
+        return 1;
+      }
+      return 0;
+    } else if (right == INFINITY && abs(left) > 1) {
+      ARROW_PREDICT_FALSE(MultiplyWithOverflow(2, std::numeric_limits<Arg0>::max(), &result));
+      return result;
+    } else if (right == -INFINITY && abs(left) > 1) {
+      return 0;
+    }
```

Comment:
I don't see why it's so complicated. This function is only for integer power, all args are supposed to be integers (see checked version). So there will be no things like `infinity`, which is only for floating points. I think just verify that the exponent is non-negtive should be enough.

Reply:
Good point. Fixed.

Reply:
I've removed quite some logic, but maybe more still more can be removed.